### DISCUSSION TITLE
Consider faster compression algorithm

### DISCUSSION THREAD

**Discussion by anayrat** [apache/airflow#26478] [Ideas]
### Description

Hello,

I suggest to consider faster compression algorithm, such as zstd or lz4.

### Use case/motivation

Such algorithms are way faster than gzip and as efficient as gzip : 
https://github.com/facebook/zstd#benchmarks

Furthermore, the cpu cost will be lowered.

Maybe you can consider slower but more efficient algorithm, when size matter more than time to compress/decompress (lzma).

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
You have not written where to consider it, so it's hard to say what you propose.

**Reply by potiuk** [reply]
And BTW. When/if you propose some details, do take into account context of where you want to propose a replacement.

There are multiple reasons one can select various compression formats other than speed of compression or size (for example availability on multiple platforms, applicability to compress the kind of data we want to compress, availability of tools among the users, backwards compatibility, historical choices, etc.etc. 

So when you propose some details please look thoroughly at all the factors and present a hollistic approach taking into accounts those.

**Reply by anayrat** [reply]
One of my customer use Postgres operator and copy the table csv with gzip compression to send them to bigquery.
I think, this step could be faster with faster algorithm. I suppose several operators" can benefit of such algorithm.
Furthermore, I am not an airflow user thought.

It is just an idea of improvements more than a real necessity.

**Reply by potiuk** [reply]
There is the saying "premature optimisation is a root of all evil". 

First let's find what needs to be optimized and why, and whether it makes sense at all. 

Going the other way round is a recipe for troubles. if you want to propose "a chnage" - I suggest you first research where and how rather than make a "generic" proposal that is not actionable without precisely that - doing a research. Without the research, it's just noise.

**Reply by anayrat** [reply]
I already answered to "where" : my customer use Postgres operator.
I can't answer to how, I am not an airflow developer, I just wanted to report a possible improvement.

> "premature optimisation is a root of all evil".

To give you a few numbers. We have a job which fetch a big table with a server side cursor which size is 600K record.
With gzip, it takes 50-60s to fetch records for each iteration. Without gzip, it takes 30-40s.

**Reply by potiuk** [reply]
Airflow has almost  2200 contributors. Mostly users like you who (when they saw a way to improve things) decided to pay back for the free software they use and contributed improvements. This is how Open-source works. Airflow is not created by a "few Airflow developers" but many, many users and contributors create it.

Since you already explained that you have a customer and your customer has some way of running it and you tested it and since you seem that you are able to justify and test it, it only sounds reasonable that you probably have enough skills to contribute a PR.

But If you think you cannot do it, feel free to open a feature request where (in the body rather than in the comment) you specify where you think things can be improved. Feel absolutely free to do it as long as you explain what you are after (Postgres Operator and sending it to bigquery). 

But  - let's be clear - creating a feature request, does not work the way you think it works (likely). In Open Source project like this we - maintainers - can mark such an issue as "good first issue" (assumiing that it iwll be properly explained what it is all about) and if somoene (one of those 2200 people) find an interest in implementing it and lead it to completion - only then it will be implemented. Not sooner. There is no "magic comittee" that decides which features will get implemented and which not. Features get implemented when ... someone implements it. So if you really want to get things to be implemented, your best bet is to actually implement it, otherwise you will need to wait until someone does it and you have no influence nor ways to make ithappen(other than advocating for your feature here or in the slack or devlist). And if you implement it yourself in a PR, you don't even need to create a feature request - PR is enough. So eventually implementing something might simply be even faster than properly explaining it in a feature request (as you see the history here - you need to spend more than one vague sentence when you describe such request and the discussion it sparked probably took you more time and energy already than implementing it). 

This is how open source works.

**Reply by anayrat** [reply]
Wow, that's pretty rude!
I was just suggesting to think about other algorithm. I am not a python developer, so I can't contribute to your project.
Sometimes, open source is considering ideas from others even if they are not capable of writing them by themselves.

FYI, I already contributed to opensource project : https://git.postgresql.org/gitweb/?p=postgresql.git&a=search&h=HEAD&st=commit&s=nayrat
So, you don't have to be condescending and tel me "how open source works"

I won't bother you again.

**Reply by potiuk** [reply]
This was just about setting expectations. Not everyone knows it, and if you do now, then likely you also know that when you propose something, it is good to explain what you propose, rather than vague "improve compression" without any details. I thought that someone who knows how open source works, also knows that they shoudl describe more details when they open such a proposal - so I assumed I have to explain it. But apparently I was wrong, and someone might know how open-source works, yet fail to make well explained issue. 

The world is full of surprises, every day.
